doesn't detect LXD even with snap lxd 2.18 installed

## Report

Ubuntu 16.04

Button is disabled in UI saying that LXD is not available.

Log:
```
2017-10-14 02:32:16,203 [DEBUG] conjure-up/kubernetes-core - provider.py:371 - LXD query cmd: lxc query --wait -X GET /1.0
2017-10-14 02:32:16,219 [ERROR] conjure-up/kubernetes-core - provider.py:376 - Unable to parse JSON output from LXD
```
Running the command with the same user in bash works.
```
lxc query --wait -X GET /1.0
{
	"api_extensions": [
		"storage_zfs_remove_snapshots",
		"container_host_shutdown_timeo
...
```
```
snap list
Name        Version    Rev   Developer  Notes
conjure-up  2.3.1      745   canonical  classic
core        16-2.28.1  3017  canonical  core
lxd         2.18       4482  canonical  -
```

--- a/conjureup/app.py
+++ b/conjureup/app.py
@@ -371,6 +371,7 @@ def main():
         AddonModel.load_spell_addons()
 
     app.env['CONJURE_UP_CACHEDIR'] = app.argv.cache_dir
+    app.env['PATH'] = "/snap/bin:{}".format(app.env['PATH'])
 
     if app.argv.show_env:
         if app.endpoint_type in [None, EndpointType.LOCAL_SEARCH]:

--- a/conjureup/controllers/clouds/common.py
+++ b/conjureup/controllers/clouds/common.py
@@ -17,15 +17,14 @@ class BaseCloudController:
 
         while not self.cancel_monitor.is_set():
             try:
+                provider._set_lxd_dir_env()
                 compatible = await provider.is_server_compatible()
                 if compatible:
                     events.LXDAvailable.set()
                     self.cancel_monitor.set()
                     cb()
                     return
-            except (LocalhostError, LocalhostJSONError):
-                provider._set_lxd_dir_env()
-            except FileNotFoundError:
+            except (LocalhostError, LocalhostJSONError, FileNotFoundError):
                 pass
             await run_with_interrupt(asyncio.sleep(2),
                                      self.cancel_monitor)

--- a/conjureup/controllers/clouds/common.py
+++ b/conjureup/controllers/clouds/common.py
@@ -17,6 +17,11 @@ class BaseCloudController:
 
         while not self.cancel_monitor.is_set():
             try:
+                compatible = await provider.is_server_compatible()
+                if not compatible:
+                    return await run_with_interrupt(asyncio.sleep(2),
+                                                    self.cancel_monitor)
+
                 await provider.is_server_available()
                 events.LXDAvailable.set()
                 self.cancel_monitor.set()

--- a/conjureup/controllers/clouds/common.py
+++ b/conjureup/controllers/clouds/common.py
@@ -19,13 +19,14 @@ class BaseCloudController:
             try:
                 compatible = await provider.is_server_compatible()
                 if not compatible:
-                    return await run_with_interrupt(asyncio.sleep(2),
-                                                    self.cancel_monitor)
-
-                await provider.is_server_available()
-                events.LXDAvailable.set()
-                self.cancel_monitor.set()
-                cb()
+                    await run_with_interrupt(asyncio.sleep(2),
+                                             self.cancel_monitor)
+                else:
+                    await provider.is_server_available()
+                    events.LXDAvailable.set()
+                    self.cancel_monitor.set()
+                    cb()
+                    return
             except (LocalhostError, LocalhostJSONError):
                 provider._set_lxd_dir_env()
                 await run_with_interrupt(asyncio.sleep(2),

--- a/conjureup/controllers/clouds/common.py
+++ b/conjureup/controllers/clouds/common.py
@@ -18,8 +18,9 @@ class BaseCloudController:
         while not self.cancel_monitor.is_set():
             try:
                 provider._set_lxd_dir_env()
-                compatible = await provider.is_server_compatible()
-                if compatible:
+                client_compatible = await provider.is_client_compatible()
+                server_compatible = await provider.is_server_compatible()
+                if client_compatible and server_compatible:
                     events.LXDAvailable.set()
                     self.cancel_monitor.set()
                     cb()

--- a/conjureup/controllers/clouds/common.py
+++ b/conjureup/controllers/clouds/common.py
@@ -18,19 +18,14 @@ class BaseCloudController:
         while not self.cancel_monitor.is_set():
             try:
                 compatible = await provider.is_server_compatible()
-                if not compatible:
-                    await run_with_interrupt(asyncio.sleep(2),
-                                             self.cancel_monitor)
-                else:
-                    await provider.is_server_available()
+                if compatible:
                     events.LXDAvailable.set()
                     self.cancel_monitor.set()
                     cb()
                     return
             except (LocalhostError, LocalhostJSONError):
                 provider._set_lxd_dir_env()
-                await run_with_interrupt(asyncio.sleep(2),
-                                         self.cancel_monitor)
             except FileNotFoundError:
-                await run_with_interrupt(asyncio.sleep(5),
-                                         self.cancel_monitor)
+                pass
+            await run_with_interrupt(asyncio.sleep(2),
+                                     self.cancel_monitor)

--- a/conjureup/controllers/clouds/tui.py
+++ b/conjureup/controllers/clouds/tui.py
@@ -1,4 +1,4 @@
-from conjureup import controllers, juju, utils
+from conjureup import controllers, events, juju, utils
 from conjureup.app_config import app
 from conjureup.consts import cloud_types
 
@@ -17,15 +17,25 @@ class CloudsController(BaseCloudController):
 
         return controllers.use('credentials').render()
 
-    def render(self):
+    async def _check_lxd_compat(self):
         utils.info(
             "Summoning {} to {}".format(app.argv.spell, app.provider.cloud))
         if app.provider.cloud_type == cloud_types.LOCALHOST:
+
             try:
                 app.provider._set_lxd_dir_env()
+                client_compatible = await app.provider.is_client_compatible()
+                server_compatible = await app.provider.is_server_compatible()
+                if client_compatible and server_compatible:
+                    self.finish()
+                else:
+                    utils.error("LXD Server or LXC client not compatible")
+                    events.Shutdown.set(1)
             except app.provider.LocalhostError:
                 raise
-        self.finish()
+
+    def render(self):
+        app.loop.create_task(self._check_lxd_compat())
 
 
 _controller_class = CloudsController

--- a/conjureup/controllers/clouds/tui.py
+++ b/conjureup/controllers/clouds/tui.py
@@ -1,6 +1,4 @@
-from pathlib import Path
-
-from conjureup import controllers, events, juju, utils
+from conjureup import controllers, juju, utils
 from conjureup.app_config import app
 from conjureup.consts import cloud_types
 
@@ -23,10 +21,10 @@ class CloudsController(BaseCloudController):
         utils.info(
             "Summoning {} to {}".format(app.argv.spell, app.provider.cloud))
         if app.provider.cloud_type == cloud_types.LOCALHOST:
-            if not Path('/snap/bin/lxc').exists():
-                utils.error("Unable to find lxc executable, please make "
-                            "sure LXD is installed: `sudo snap install lxd`")
-                events.Shutdown.set(1)
+            try:
+                app.provider._set_lxd_dir_env()
+            except app.provider.LocalhostError:
+                raise
         self.finish()
 
 

--- a/conjureup/models/provider.py
+++ b/conjureup/models/provider.py
@@ -426,19 +426,21 @@ class Localhost(BaseProvider):
                 _pools.move_to_end('default', last=False)
         return _pools
 
-    async def is_server_available(self):
+    async def is_server_compatible(self):
         """ Waits and checks if LXD server becomes available
 
         We'll want to loop here and ignore most errors until have retries have
         been met
         """
         try:
-            return await self.query()
+            out = await self.query()
+            server_ver = out['environment']['server_version']
+            return parse_version(server_ver) >= self.minimum_support_version
         except (LocalhostError, LocalhostJSONError, FileNotFoundError):
-            raise
+            return False
 
-    async def is_server_compatible(self):
-        """ Checks if LXD server version is compatible with conjure-up
+    async def is_client_compatible(self):
+        """ Checks if LXC version is compatible with conjure-up
         """
         try:
             _, out, err = await utils.arun([self.lxc_bin, 'version'])

--- a/conjureup/models/provider.py
+++ b/conjureup/models/provider.py
@@ -356,8 +356,8 @@ class Localhost(BaseProvider):
             app.env['LXD_DIR'] = str(self.lxd_socket_dir)
         else:
             raise LocalhostError(
-                "Unable to find /snap/bin/lxd. Make sure `snap list` "
-                "shows lxd as installed, otherwise, run `sudo snap "
+                "Unable to find /snap/bin/lxd. Make sure `snap info lxd` "
+                "shows as installed, otherwise, run `sudo snap "
                 "install lxd` and restart conjure-up.")
 
     async def query(self, segment='', method="GET"):

--- a/conjureup/models/provider.py
+++ b/conjureup/models/provider.py
@@ -346,6 +346,7 @@ class Localhost(BaseProvider):
         self.network_interface = None
         self.minimum_support_version = parse_version('2.17')
         self.available = False
+        self.lxc_bin = None
         self._set_lxd_dir_env()
 
     def _set_lxd_dir_env(self):
@@ -354,11 +355,18 @@ class Localhost(BaseProvider):
         if Path('/snap/bin/lxd').exists():
             self.lxd_socket_dir = Path('/var/snap/lxd/common/lxd')
             app.env['LXD_DIR'] = str(self.lxd_socket_dir)
+            self.lxc_bin = '/snap/bin/lxc'
+        elif Path('/usr/bin/lxd').exists():
+            self.lxd_socket_dir = Path('/var/lib/lxd')
+            app.env['LXD_DIR'] = str(self.lxd_socket_dir)
+            self.lxc_bin = '/usr/bin/lxc'
         else:
             raise LocalhostError(
-                "Unable to find /snap/bin/lxd. Make sure `snap info lxd` "
+                "Unable to find a lxd binary. Make sure `snap info lxd` "
                 "shows as installed, otherwise, run `sudo snap "
                 "install lxd` and restart conjure-up.")
+        app.log.debug("LXD environment set: binary {} lxd_dir {}".format(
+            self.lxc_bin, self.lxd_socket_dir))
 
     async def query(self, segment='', method="GET"):
         """ Query lxc api server
@@ -371,15 +379,15 @@ class Localhost(BaseProvider):
             segment_prefix = Path('/1.0')
             url = str(segment_prefix / segment)
         try:
-            cmd = ['/snap/bin/lxc', 'query', '--wait', '-X', method.upper(),
+            cmd = [self.lxc_bin, 'query', '--wait', '-X', method.upper(),
                    url]
             app.log.debug("LXD query cmd: {}".format(" ".join(cmd)))
             _, out, err = await utils.arun(cmd)
             return json.loads(out)
         except json.decoder.JSONDecodeError:
             err = ("Unable to parse JSON output from LXD, does "
-                   "`/snap/bin/lxc query --wait -X GET /1.0` "
-                   "return info about the LXD server?")
+                   "`{} query --wait -X GET /1.0` "
+                   "return info about the LXD server?".format(self.lxc_bin))
             app.log.error(err)
             raise LocalhostJSONError(err)
         except FileNotFoundError as e:
@@ -433,11 +441,12 @@ class Localhost(BaseProvider):
         """ Checks if LXD server version is compatible with conjure-up
         """
         try:
-            out = await self.query()
-        except LocalhostError:
-            return False
-        server_ver = out['environment']['server_version']
-        return parse_version(server_ver) <= self.minimum_support_version
+            _, out, err = await utils.arun([self.lxc_bin, 'version'])
+            server_ver = out.strip()
+            return parse_version(server_ver) >= self.minimum_support_version
+        except CalledProcessError as e:
+            app.log.error(e)
+            raise LocalhostError(e)
 
 
 class Azure(BaseProvider):

--- a/conjureup/models/provider.py
+++ b/conjureup/models/provider.py
@@ -354,6 +354,11 @@ class Localhost(BaseProvider):
         if Path('/snap/bin/lxd').exists():
             self.lxd_socket_dir = Path('/var/snap/lxd/common/lxd')
             app.env['LXD_DIR'] = str(self.lxd_socket_dir)
+        else:
+            raise LocalhostError(
+                "Unable to find /snap/bin/lxd. Make sure `snap list` "
+                "shows lxd as installed, otherwise, run `sudo snap "
+                "install lxd` and restart conjure-up.")
 
     async def query(self, segment='', method="GET"):
         """ Query lxc api server
@@ -366,13 +371,15 @@ class Localhost(BaseProvider):
             segment_prefix = Path('/1.0')
             url = str(segment_prefix / segment)
         try:
-            cmd = ['lxc', 'query', '--wait', '-X', method.upper(),
+            cmd = ['/snap/bin/lxc', 'query', '--wait', '-X', method.upper(),
                    url]
             app.log.debug("LXD query cmd: {}".format(" ".join(cmd)))
             _, out, err = await utils.arun(cmd)
             return json.loads(out)
         except json.decoder.JSONDecodeError:
-            err = "Unable to parse JSON output from LXD"
+            err = ("Unable to parse JSON output from LXD, does "
+                   "`/snap/bin/lxc query --wait -X GET /1.0` "
+                   "return info about the LXD server?")
             app.log.error(err)
             raise LocalhostJSONError(err)
         except FileNotFoundError as e:

--- a/conjureup/ui/views/cloud.py
+++ b/conjureup/ui/views/cloud.py
@@ -146,6 +146,7 @@ class CloudView(WidgetWrap):
                                     "for this message to disappear:\n\n"
                                     "  $ sudo snap install lxd\n"
                                     "  $ sudo usermod -a -G lxd <youruser>\n"
+                                    "  $ newgrp lxd\n"
                                     "  $ /snap/bin/lxd init --auto\n"
                                     "  $ /snap/bin/lxc network create lxdbr0 "
                                     "ipv4.address=auto ipv4.nat=true "

--- a/conjureup/ui/views/cloud.py
+++ b/conjureup/ui/views/cloud.py
@@ -145,6 +145,7 @@ class CloudView(WidgetWrap):
                                     "LXD not found, please install and wait "
                                     "for this message to disappear:\n\n"
                                     "  $ sudo snap install lxd\n"
+                                    "  $ sudo usermod -a -G lxd <youruser>\n"
                                     "  $ /snap/bin/lxd init --auto\n"
                                     "  $ /snap/bin/lxc network create lxdbr0 "
                                     "ipv4.address=auto ipv4.nat=true "

--- a/snap/wrappers/juju
+++ b/snap/wrappers/juju
@@ -7,11 +7,11 @@ snap_lxd_socket=/var/snap/lxd/common/lxd
 
 lxd_binary=$(which lxd)
 
-# Prefer debian installed lxd, if availible
-if [ "$lxd_binary" = "/usr/bin/lxd" ]; then
-    export LXD_DIR=$debian_socket
-elif [ "$lxd_binary" = "/snap/bin/lxd" ]; then
+# Prefer snap installed lxd, if availible
+if [ "$lxd_binary" = "/snap/bin/lxd" ]; then
     export LXD_DIR=$snap_lxd_socket
+elif [ "$lxd_binary" = "/usr/bin/lxd" ]; then
+    export LXD_DIR=$debian_socket
 fi
 export PATH=$SNAP/bin:$SNAP/usr/bin:/snap/bin:$PATH
 export LD_LIBRARY_PATH

--- a/snap/wrappers/juju
+++ b/snap/wrappers/juju
@@ -5,6 +5,7 @@
 debian_socket=/var/lib/lxd
 snap_lxd_socket=/var/snap/lxd/common/lxd
 
+export PATH=$SNAP/bin:$SNAP/usr/bin:/snap/bin:$PATH
 lxd_binary=$(which lxd)
 
 # Prefer snap installed lxd, if availible
@@ -13,7 +14,6 @@ if [ "$lxd_binary" = "/snap/bin/lxd" ]; then
 elif [ "$lxd_binary" = "/usr/bin/lxd" ]; then
     export LXD_DIR=$debian_socket
 fi
-export PATH=$SNAP/bin:$SNAP/usr/bin:/snap/bin:$PATH
 export LD_LIBRARY_PATH
 LD_LIBRARY_PATH=$SNAP/lib:$SNAP/usr/lib/$(uname -p)-linux-gnu/
 

--- a/test/test_controllers_clouds_tui.py
+++ b/test/test_controllers_clouds_tui.py
@@ -47,11 +47,12 @@ class CloudsTUIRenderTestCase(unittest.TestCase):
 
     def test_render(self):
         "Rendering with a known cloud should call finish"
+        self.controller._check_lxd_compat = MagicMock()
         self.mock_app.provider.cloud = "aws"
         t = ['aws']
         self.mock_juju.get_clouds.return_value.keys.return_value = t
         self.controller.render()
-        self.mock_finish.assert_called_once_with()
+        assert self.mock_app.loop.create_task.called
 
 
 class CloudsTUIFinishTestCase(unittest.TestCase):

